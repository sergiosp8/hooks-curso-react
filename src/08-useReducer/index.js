export * from './TodoAdd'
export * from './TodoApp'
export * from './TodoList'
export * from './intro-reducer'
export * from './todoReducer'
