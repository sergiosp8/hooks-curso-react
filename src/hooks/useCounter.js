import { useState } from 'react'

export const useCounter = (initialValue = 10) => {
  const [counter, setCounter] = useState(initialValue)

  const increment = (value = 1) => {
    setCounter((prev) => prev + value)
  }

  const decrement = (value = 1) => {
    // if ( counter === 0 ) return;

    setCounter((prev) => prev - value)
  }

  const reset = () => {
    setCounter(initialValue)
  }

  return {
    counter,
    increment,
    decrement,
    reset
  }
}
