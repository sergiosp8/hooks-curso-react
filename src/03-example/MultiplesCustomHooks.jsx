import { useCounter, useFetch } from '../hooks'
import { LoadingQuote, Quote } from './components'

const MultiplesCustomHooks = () => {
  const { counter, increment } = useCounter(1)
  const { data, isLoading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)

  const { author, quote } = !!data && data[0]

  return (
    <>
      <h1>Breaking Bad Quotes</h1>
      <hr />

      {isLoading ? <LoadingQuote /> : <Quote quote={quote} author={author} />}

      <button onClick={() => increment()} className="btn btn-primary">
        Next quote
      </button>
    </>
  )
}

export default MultiplesCustomHooks
