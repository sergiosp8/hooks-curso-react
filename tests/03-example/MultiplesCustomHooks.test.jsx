import { fireEvent, render, screen } from '@testing-library/react'
import MultiplesCustomHooks from '../../src/03-example/MultiplesCustomHooks'
import { useCounter, useFetch } from '../../src/hooks'

jest.mock('../../src/hooks/useFetch')
jest.mock('../../src/hooks/useCounter')

describe('Pruebas en <MultiplesCustomHooks/>', () => {
  const mockIncrement = jest.fn()
  useCounter.mockReturnValue({
    counter: 1,
    increment: mockIncrement
  })

  beforeEach(() => {
    jest.clearAllMocks()
  })
  test('el componente debe ser igual al snapshot', () => {
    useFetch.mockReturnValue({ data: null, isLoading: true, hasError: null })

    const container = render(<MultiplesCustomHooks />)
    expect(container).toMatchSnapshot()
  })

  test('debe de mostrar el componente por defecto', () => {
    render(<MultiplesCustomHooks />)

    expect(screen.getByText('Loading...'))
    expect(screen.getByText('Breaking Bad Quotes'))
    // screen.debug()
  })

  test('debe de mostrar un Quote', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'sergio',
          quote: 'hola'
        }
      ],
      isLoading: false,
      hasError: null
    })
    render(<MultiplesCustomHooks />)
    expect(screen.getByText('hola')).toBeTruthy()
    expect(screen.getByText('sergio')).toBeTruthy()

    // screen.debug()
  })

  test('debde de llamar la funcion increment', () => {
    render(<MultiplesCustomHooks />)
    const nextButton = screen.getByRole('button')
    fireEvent.click(nextButton)
    expect(mockIncrement).toHaveBeenCalled()
    // screen.debug()
  })
})
