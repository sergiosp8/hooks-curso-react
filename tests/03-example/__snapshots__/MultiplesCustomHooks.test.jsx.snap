// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultiplesCustomHooks/> el componente debe ser igual al snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Breaking Bad Quotes
      </h1>
      <hr />
      <div
        class="alert alert-info text-center"
      >
        Loading...
      </div>
      <button
        class="btn btn-primary"
      >
        Next quote
      </button>
    </div>
  </body>,
  "container": <div>
    <h1>
      Breaking Bad Quotes
    </h1>
    <hr />
    <div
      class="alert alert-info text-center"
    >
      Loading...
    </div>
    <button
      class="btn btn-primary"
    >
      Next quote
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
